reasoner/planner/ConjunctionSummarizer.java

```diff
@@ -0,0 +1,232 @@
+    }
+
+    public Set<ResolvableConjunction> dependencies(Concludable concludable) {
+        return iterate(logicMgr.applicableRules(concludable).keySet()).map(rule -> rule.condition().conjunction()).toSet();
+    }
+
+    public ConjunctionSummary conjunctionSummary(ResolvableConjunction conjunction) {
+        if (!conjunctionSummaries.containsKey(conjunction)) {
+            recursivelySummarizeConjunctions(conjunction);
+        }
+        return conjunctionSummaries.get(conjunction);
+    }
+
+    private ConjunctionSummary createConjunctionSummary(ResolvableConjunction conjunction, Set<Pair<Concludable, ResolvableConjunction>> cyclicConcludables) {
+        Set<Resolvable<?>> resolvables = logicMgr.compile(conjunction);
+        Map<Concludable, Set<ResolvableConjunction>> cyclicDependencies = new HashMap<>();
+        Map<Concludable, Set<ResolvableConjunction>> acyclicDependencies = new HashMap<>();
+        iterate(resolvables).filter(Resolvable::isConcludable).map(Resolvable::asConcludable).forEachRemaining(concludable -> {
+            cyclicDependencies.put(concludable, new HashSet<>());
+            acyclicDependencies.put(concludable, new HashSet<>());
+        });
+
+        iterate(cyclicConcludables).forEachRemaining(concludableDependency -> {
+            cyclicDependencies.get(concludableDependency.first()).add(concludableDependency.second());
+        });
+
+        iterate(resolvables).filter(Resolvable::isConcludable).map(Resolvable::asConcludable).forEachRemaining(concludable -> {
+            iterate(dependencies(concludable))
+                    .filter(dependency -> !cyclicDependencies.get(concludable).contains(dependency))
+                    .forEachRemaining(dependency -> acyclicDependencies.get(concludable).add(dependency));
+        });
+
+        return new ConjunctionSummary(conjunction, resolvables, cyclicDependencies, acyclicDependencies);
+    }
+
+    private void recursivelySummarizeConjunctions(ResolvableConjunction root) {
+        ConjunctionConcludableStack stack = new ConjunctionConcludableStack();
+        Map<ResolvableConjunction, Set<Pair<Concludable, ResolvableConjunction>>> cyclicConcludables = new HashMap<>();
+        findCyclicConcludables(root, stack, cyclicConcludables);
+        iterate(stack.visited()).forEachRemaining(conjunction -> {
+            assert !conjunctionSummaries.containsKey(conjunction);
+            conjunctionSummaries.put(conjunction, createConjunctionSummary(conjunction, cyclicConcludables.getOrDefault(conjunction, new HashSet<>())));
+        });
+    }
+
+    private void findCyclicConcludables(ResolvableConjunction conjunction, ConjunctionConcludableStack stack, Map<ResolvableConjunction, Set<Pair<Concludable, ResolvableConjunction>>> cyclicConcludables) {
+        if (conjunctionSummaries.containsKey(conjunction)) {
+            return; // We only add conjunction summaries at the end of a search -> this does not loop back.
+        }
```

Comment:
Optimisation added: We start our search at a root which is not part of a cycle. We summarise the conjunctions in a subgraph AFTER the search. This ensures all cyclic dependencies in the subgraph have been found, and we don't have to recurse down this path from other entry points.